common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java

```diff
@@ -0,0 +1,166 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.embeddedwebview;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.net.http.SslError;
+import android.view.View;
+import android.webkit.ClientCertRequest;
+import android.webkit.HttpAuthHandler;
+import android.webkit.SslErrorHandler;
+import android.webkit.WebResourceError;
+import android.webkit.WebResourceRequest;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationRequest;
+import com.microsoft.identity.common.internal.util.StringUtil;
+
+public abstract class OAuth2WebViewClient extends WebViewClient {
+    /**
+     * Redirect URLs are a critical part of the OAuth flow. After a user successfully authorizes an
+     * application, the authorization server will redirect the user back to the application with
+     * either an authorization code or access token in the URL.
+     *
+     * @TODO Will it be a part of the auth configuration class?
+     */
+    private final String mRedirectURL;
+
+    private final AuthorizationRequest mRequest; //NOPMD
+
+    private final Context mContext; //NOPMD
+
+    static final String BLANK_PAGE = "about:blank";
+
+    abstract void showSpinner(boolean showing);
+
+    OAuth2WebViewClient(final Context context, final String redirectURL, final AuthorizationRequest request) {
+        //the validation of redirect url and authorization request should be in upper level before launching the webview.
+        if (null == context || null == request || StringUtil.isEmpty(redirectURL)) {
+            throw new IllegalArgumentException("Null parameter to initialize OAuth2WebViewClient.");
+        }
+
+        mContext = context;
+        mRedirectURL = redirectURL;
+        mRequest = request;
+    }
+
+    public String getmRedirectURL() {
+        return mRedirectURL;
+    }
+
+    /**
+     * Notify the host application that a page has started loading.
+     * @param view The WebView that is initiating the callback.
+     * @param url The url to be loaded.
+     * @param favicon The favicon for this page if it already exists in the
+     *            database.
+     */
+    @Override
+    public void onPageStarted(WebView view, String url, Bitmap favicon) {

```

Comment:
Why do we have our own javadoc for an empty override?

Reply:
`onPageFinished()` calls `showSpinner(false)` -- shouldn't this conversely call `showSpinner(true)`?